elixir/apps/api/lib/api/client/channel.ex

```diff
@@ -130,45 +143,102 @@ defmodule API.Client.Channel do
+  def handle_info({:delete_resource, _resource_id}, socket) do
+    {:noreply, socket}
+  end
+
+  # Those events are broadcasted by Actors whenever a membership is created or deleted
+  def handle_info({:create_membership, _actor_id, group_id}, socket) do
+    :ok = Policies.subscribe_for_events_for_actor_group(group_id)
+    {:noreply, socket}
+  end
+
+  def handle_info({:delete_membership, _actor_id, group_id}, socket) do
+    :ok = Policies.unsubscribe_from_events_for_actor_group(group_id)
+    {:noreply, socket}
+  end
+
+  # This message is received when there is a policy created or enabled
+  # allowing access to a resource by a client actor group
+  def handle_info({:allow_access, policy_id, actor_group_id, resource_id}, socket) do
     OpenTelemetry.Ctx.attach(socket.assigns.opentelemetry_ctx)
     OpenTelemetry.Tracer.set_current_span(socket.assigns.opentelemetry_span_ctx)
 
-    OpenTelemetry.Tracer.with_span "client.resource_updated",
-      attributes: %{resource_id: resource_id} do
-      with {:ok, resource} <- Resources.fetch_resource_by_id(resource_id, socket.assigns.subject) do
-        push(socket, "resource_updated", Views.Resource.render(resource))
+    OpenTelemetry.Tracer.with_span "client.allow_access",
+      attributes: %{
+        policy_id: policy_id,
+        actor_group_id: actor_group_id,
+        resource_id: resource_id
+      } do
+      :ok = Resources.subscribe_for_events_for_resource(resource_id)
+
+      case Resources.fetch_and_authorize_resource_by_id(resource_id, socket.assigns.subject) do
+        {:ok, resource} ->
+          push(socket, "resource_created_or_updated", Views.Resource.render(resource))
+
+        {:error, _reason} ->
+          :ok
       end
 
       {:noreply, socket}
     end
   end
 
-  def handle_info({:resource_removed, resource_id}, socket) do
+  # This message is received when the policy
+  # allowing access to a resource by a client actor group is deleted
+  def handle_info({:reject_access, policy_id, actor_group_id, resource_id}, socket) do
     OpenTelemetry.Ctx.attach(socket.assigns.opentelemetry_ctx)
     OpenTelemetry.Tracer.set_current_span(socket.assigns.opentelemetry_span_ctx)
 
-    OpenTelemetry.Tracer.with_span "client.resource_removed",
-      attributes: %{resource_id: resource_id} do
-      push(socket, "resource_removed", resource_id)
+    OpenTelemetry.Tracer.with_span "client.reject_access",
+      attributes: %{
+        policy_id: policy_id,
+        actor_group_id: actor_group_id,
+        resource_id: resource_id
+      } do
+      :ok = Resources.unsubscribe_from_events_for_resource(resource_id)
+
+      # We potentially can re-create the flow but this will require keep tracking of client connections to gateways,
+      # which is not worth it as this case should be pretty rare. Instead we just tell client to remove it
+      # and the recreate it right away if there is another allowing access to it.
+      push(socket, "resource_deleted", resource_id)
```

Comment:
@conectado here there is a handling of an edge case where somebody disables/deletes a policy that the client was using but there is a different one that will allow the same access. We will push `resource_deleted` message first and then bring it back via `resource_created_or_updated` if there is a different policy that allows access to it, this will happen rather quickly (one DB roundtrip latency between messages).

Reply:
This will interrupt flows between clients and gateways.

We could buffer messages and dispatch them after a small delay to prevent this, but it'll be not trivial to implement.

Reply:
We should interrupt the flow because we need a whole new authorization cycle to happen in order for everything to work in predictable manner.

Otherwise you will see a flow in audit log for a deleted policy that is expired but the traffic still will flow..